### ISSUE TITLE
Offloading the Hosting and Maintenance of the Non-English Docs

### ISSUE THREAD

**Issue by WestLangley** [mrdoob/three.js#24984]
I propose that the Non-English docs be hosted elsewhere -- and maintained by others.

Being responsible for ensuring the English and non-English docs remain in sync is not an efficient use of the three.js maintainers' time.

Furthermore, it is a burden that is only going to increase as more languages are supported and the docs become more complete.

This thread is for discussing how offloading the hosting and maintenance of the Non-English docs may best be accomplished.

I would be helpful if someone would be willing to take responsibility for this task. 😇

**Comment by takahirox** [comment]
+1 for this

> Being responsible for ensuring the English and non-English docs remain in sync is not an efficient use of the three.js maintainers' time.

I agree with this. I have been feeling the same.

**Comment by Mugen87** [comment]
> Being responsible for ensuring the English and non-English docs remain in sync

I think it's sufficient if we just state that the maintainers only manage the English docs from now on. In this way, we can still keep the other languages which makes it easier to overlook the translation process and host everything.

**Comment by Mugen87** [comment]
BTW: This discussion should also include the translations of the manual.

**Comment by WestLangley** [comment]
>In this way, we can still keep the other languages in the repository.

No. That would be a poor decision from a management standpoint. The maintainers would then still be responsible for responding to bug reports and merge requests for the other languages.

**Comment by Mugen87** [comment]
The maintenance in that regard is minor. I have no problem to be responsible for this task.

**Comment by donmccurdy** [comment]
Thoughts:

1. For a mature project like three.js, high-quality translations (or internationalization / "i18n") feel very important, and I'd be sorry for that area of the project to become stale. Really grateful for the folks doing these translations.
2. If we simply offload translation to "the community" and don't set up a good workflow for keeping things in sync, and I expect the translations may degrade quickly. Our hands-off approach to `@types/three` will not suffice here.
3. With @Mugen87, I am glad to be involved in this task in any way. This is important to me.
4. It may be that the current setup is not that great for translators either, and offloading (done well) might actually improve things. For example, a separate repository might allow maintainers of the translations to set up tooling that we haven't provided here, for quickly identifying:
    - new strings, not yet translated to language X
    - updated strings, changed since translation to language X)
    - progress, % of documentation translated to language X
6. I'd love to hear from people doing these translations: What works today? What doesn't? How do you feel having translations occur in a separate repository, possibly with more dedicated tooling to support that? Are there good tools for this you've used before?

**Comment by WestLangley** [comment]
>I'd be sorry for that area of the project to become stale.

They are already stale.

**Comment by Mugen87** [comment]
The maintenance work I do is straightforward right now because when a single documentation page is updated, the changes can often be copy/pasted with an editor (new code example, updated signature etc.). But granted it is not a scalable solution to copy each documentation page and translate the contents by hand. However, using a real translation tool with placeholders in the documentation templates and a language database seems out of scope.

I'm not an expert in this field but would it be possible to avoid any manual translation and just ask the developer to use a translation tool e.g. like Google Translate? Are the result considered to be good enough now?

I've tested this today with some guides from the documentation and the results are about right (at least when you translate into German). I mean the tool translates some words that it shouldn't (like render target) and the structure of a sentence is sometimes a bit weird but at least you understand the actual content of the article.

I wonder if there is a way to improve the English docs so a translation tool can produce better results. Like preventing the translation of certain words or defining their syntax (e.g. that "render targets" should not be translated and is the plural of "render target"). Maybe it would be possible to tweak the results up to a point where it can be considered as "good enough" and a manual translation isn't necessary anymore.

**Comment by donmccurdy** [comment]
> However, using a real translation system with placeholders in the documentation templates and a language database seems out of scope.

I wish I knew a practical way to do this, but, yeah. I'm not sure we have the resources.

> I wonder if there is a way to improve the English docs so a translation tool can produce better results

I think one way to do this would be to change how we're using `<code>` blocks today. We use them for longform code blocks, but that's really what `<pre><code>` is for:

```html
<pre><code>
import * as THREE from 'three';

...
</code></pre>
```

If we change that, we could start using `<code/>` for inline tokens instead, like:

```html
<p>
  This guide explains how to use <code>GLTFLoader</code>.
</p>
```

Then we just need to prevent translation from affecting code tags, but limit syntax highlighting to `<pre><code>` blocks.

**Comment by Mugen87** [comment]
It seems doing this:
```
 <span translate="no">render target</span>
```
prevents Google Translate from translating the words into another language. I've tested this with Chrome and the built-in translating option. The `translate` attribute is not Google Translate specific so other tools should recognize it as well.

**Comment by Mugen87** [comment]
At least Google Translate does already not translate the code examples in the documentation. I guess this happens because since the code blocks already have the `translate="no"` attribute set.
https://github.com/mrdoob/three.js/blob/6663480c40f31629c4a0af98d5d2df7072888b45/docs/page.js#L164

**Comment by donmccurdy** [comment]
Yeah, it may be more practical to use `<code />` blocks inline, and let the script add `translate="no"`, rather than using a span with attributes each time.

**Comment by Mugen87** [comment]
I just if it's right to use the `<code />` element like that. AFAIK, it's intended to define computer code in HTML files. We just want to mark specific text from not being processed by the translator. Sounds a bit like a misusage to me. To me, the`<span>` attribute feels like the cleaner approach.

**Comment by donmccurdy** [comment]
I only mean using it when we're referring to classes, methods, and variables. It's an inline element, not a block element, in normal usage:

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/code

This helps with styling too, I'm always putting words like Group in italics to identify them, but we could use consistent styling with semantic markup instead.

But yeah, if we're trying to prevent a non-code word or phrase from being translated, then `<span translate="no" />` is probably better.

**Comment by LeviPesin** [comment]
> I'm not an expert in this field but would it be possible to avoid any manual translation and just ask the developer to use a tool e.g. like Google Translate? Are the results considered to be good enough now?

Depends on the output language -- e.g. for Russian it's generally good enough for few words or a phrase, but a paragraph can have some problems.

**Comment by Mugen87** [comment]
> But it is understandable.

This is what I'd like to clarify. Can we consider the quality of translating tools as "good enough"? And is there a way to improve the English docs such that the output is improved. 

For example when I prevent the translation of the term "render target", the German translation does not know whether the term is singular or plural. The resulting structure of a sentence can sound strange then. 

I wonder if an additional Glossary could help here: https://cloud.google.com/translate/docs/advanced/glossary

Has somebody already make experience with that technique?

**Comment by LeviPesin** [comment]
Should we actually prevent translation of terms like render targets? Maybe only proper names like algorithm names or class names (CSM, Mesh, etc) should not be translated... I'm not sure how I would translate "render target" to Russian -- both literal translation (like, "цель рендера" -- it's more of "the target of the rendering", which is wrong) and just leaving untranslated seem bad options to me. It may be more appropriate to even use transcription and "invent" a new term like "рендертаргет"...

**Comment by Mugen87** [comment]
> Should we actually prevent translation of terms like render targets? 

That was discussed above. A Glossary could also help in this context. It seems a more effective way than just using `<span translate="no" />`. Just wondering if someone already used this feature.

**Comment by mrdoob** [comment]
Google Translate has a service for translating sites but seems like it break pretty badly...

https://threejs-org.translate.goog/docs/?_x_tr_sl=en&_x_tr_tl=es&_x_tr_hl=en-US&_x_tr_pto=wapp#manual/en/introduction/Creating-a-scene

**Comment by AlexandreAllard** [comment]
> This is what I'd like to clarify. Can we consider the quality of translating tools as "good enough"? And is there a way to improve the English docs such that the output is improved.

Hello, I'm the one who started translating the documentation in french, I feel like translating tools are absolutely not "good enough". In french it creates nonsense and confuses the reader, lately one contributor translated some pages in French, after 10s I knew it was "Google Translated" because of all the weird sentences it contained.

ThreeJS has a large subset of specific words that can't be easily translated, that souldn't be translated or that need a human interpretation to be translated, so I think that the documentation is doomed to be of poor/average quality if we use translating tools. A page like the "[color management](https://threejs.org/docs/#manual/en/introduction/Color-management)" one, cannot be translated without someone picking the best/understandable sentences/words.

But on the other hand, I understand that the automated translation may be a way of having a neat translation-workflow.

**Comment by AngyDev** [comment]
Hello, I don't mind having a separate repository, if it's easier to manage, that way the translations could be maintained by those who are actually interested in doing it. I think google translate is not the best tool to automate translations, it's not always clear what it means, especially given the topic. It may not be a good idea, a system can be created that when changes are made to the English documentation an issue is opened for all other translations so that those who want to can update their chosen language.

Another translator, better than Google Translate is DeepL Translate, have you ever heard of this?

**Comment by Mugen87** [comment]
I've just mentioned Google Translate since I assume it is what Chrome is internally using when you open the context menu and use the translation option. 

Does DeepL Translate offer a service that let's you translate entire pages like in https://github.com/mrdoob/three.js/issues/24984#issuecomment-1469780634?

**Comment by NaNshekhar04** [comment]
Can I go for Hindi or Hinglish India ? @mrdoob

**Comment by AngyDev** [comment]
Yes, there are APIs that offer the translation of a document, the only thing is that there is a free plan and a pro one, with the first one, it's possible to translate until 500000 characters per month.

**Comment by Mael-Kehl** [comment]
> > This is what I'd like to clarify. Can we consider the quality of translating tools as "good enough"? And is there a way to improve the English docs such that the output is improved.
> 
> Hello, I'm the one who started translating the documentation into french, I feel like translating tools are absolutely not "good enough". In french it creates nonsense and confuses the reader, lately one contributor translated some pages in French, after 10s I knew it was "Google Translated" because of all the weird sentences it contained.
> 
> ThreeJS has a large subset of specific words that can't be easily translated, that shouldn't be translated or that need a human interpretation to be translated, so I think that the documentation is doomed to be of poor/average quality if we use translating tools. A page like the "[color management](https://threejs.org/docs/#manual/en/introduction/Color-management)" one, cannot be translated without someone picking the best/understandable sentences/words.
> 
> But on the other hand, I understand that the automated translation may be a way of having a neat translation-workflow.

Hello there, It seems like I've been hit by a lost bullet in the discussion, I used deepl on pages like geometries and I only needed to edit 20% of the results because they were pretty consistent. However in languages such as French you are obligated to always check the translator output because of the specific words you mentioned.   
Based on @AngyDev answer, I suggest using a default translator whose translation can be verified by a contributor. A check/validation Icon could be used to specify that the pages have been checked by a human, so people can adapt their reading and so have a critical opinion.

**Comment by AlexandreAllard** [comment]
> > > This is what I'd like to clarify. Can we consider the quality of translating tools as "good enough"? And is there a way to improve the English docs such that the output is improved.
> > 
> > 
> > Hello, I'm the one who started translating the documentation into french, I feel like translating tools are absolutely not "good enough". In french it creates nonsense and confuses the reader, lately one contributor translated some pages in French, after 10s I knew it was "Google Translated" because of all the weird sentences it contained.
> > ThreeJS has a large subset of specific words that can't be easily translated, that shouldn't be translated or that need a human interpretation to be translated, so I think that the documentation is doomed to be of poor/average quality if we use translating tools. A page like the "[color management](https://threejs.org/docs/#manual/en/introduction/Color-management)" one, cannot be translated without someone picking the best/understandable sentences/words.
> > But on the other hand, I understand that the automated translation may be a way of having a neat translation-workflow.
> 
> Hello there, It seems like I've been hit by a lost bullet in the discussion, I used deepl on pages like geometries and I only needed to edit 20% of the results because they were pretty consistent. However in languages such as French you are obligated to always check the translator output because of the specific words you mentioned. Based on @AngyDev answer, I suggest using a default translator whose translation can be verified by a contributor. A check/validation Icon could be used to specify that the pages have been checked by a human, so people can adapt their reading and so have a critical opinion.

Hey mate, no offense intended, no need to justify yourself or your workflow for translating the documentation, ThreeJS is only love and geometries, nothing else haha.
The only thing is that, if, whitout speaking with you, I have been able to determine that you have used an automated translation tool, it means that the output isn't natural/correct enough for it to be enjoyable for an user, that was my initial point.

**Comment by donmccurdy** [comment]
Proposal:

- [RFC: Adopting JSDoc in three.js](https://gist.github.com/donmccurdy/db75b86a28a775135e33317ce2694a32)

While I think it's a promising general direction, there is a small but important blocker described at the end.

**Comment by WSPluta** [comment]
did you see how godot is managing translation and documentation? We should create same solution for three.js 

https://hosted.weblate.org/projects/godot-engine/godot-docs/

**Comment by WestLangley** [comment]
This suggestion is not getting any traction. Closing due to lack of interest.

**Comment by Mugen87** [comment]
Please let's not close the issue. It is an important one and I highly vote for what has been suggested in https://github.com/mrdoob/three.js/issues/24984#issuecomment-1647146748.

I think it is just a matter of resources that nobody works at this issue at the moment.

**Comment by Methuselah96** [comment]
I would love to help in any way I can, primarily if it makes it easier to maintain doc comments in the TypeScript definitions, since right now I have to manually copy/paste those and reformat them into a JSDoc. I'm not sure that we can generate TypeScript definitions directly from the JSDoc, but I think it would still be useful for maintaining the TS types.

Let me know what the first step is, and I'd be happy to start working on it.

**Comment by Methuselah96** [comment]
Here's a quick proof-of-concept: https://github.com/mrdoob/three.js/pull/29357. In this case the types are simple enough that we could use the generated .d.ts without any modifications, which would be awesome.

**Comment by donmccurdy** [comment]
> I'm not sure that we can fully generate TypeScript definitions directly from the JSDoc quite yet, ...

Yes, this was the 'blocker' I mentioned in my comment – I'm not sure we could generate acceptable TypeScript definitions from JSDoc ... but I see no reason that generating _documentation_ from JSDoc shouldn't work well. If that's still enough of a win to move forward, then let's do it.

I would also be grateful for a sanity-check from one or more people who are involved in contributing translations – does contributing translated strings into a more machine-readable file (.csv, .yml, or .sqlite?) sound like an improvement over the current structure of writing in .html files? I *think* it would be, because it's much easier to identify English source strings that have yet to be translated, or that have changed since the last translation ... but I'm not translating the content myself, and I want to be sure we're making that easier not harder... 😅

**Comment by Methuselah96** [comment]
If I understand correctly, the main pain point is three.js maintainers needing to spend time keeping the non-English docs in-sync, so it seems like anything that reduces the need to update the docs in 7 different places would be an improvement.

@donmccurdy Do you have a plan for generating the English-language API docs from JSDoc? It seems like it would be nice to have that in-place before we start adding the JSDoc so that we're not adding to the duplication of docs. Or if we're fine with the duplication, I can start adding JSDoc.

**Comment by donmccurdy** [comment]
@Methuselah96 I think that's a correct statement of the pain point, yes!

If we were confident about getting TS definitions too, I had been thinking of using [ts-morph](https://ts-morph.com/) (I have some experience with this) to parse the definitions into a simple JSON representation of the three.js public API (for example, `api.json`). But there may be other/better tools for extracting a JSON representation of the API from the JSDoc, too.

Then, a script should extract all English-language strings for translation (`docs.en.csv`). Non-English language strings would appear in other files (`docs.cn.csv`). Finally, documentation would be built by a script that processes the JSON, looks up translations from the CSVs, and outputs HTML files for each language in a build step.

I think it would be good to have a proof-of-concept for the full process before we begin committing a lot of JSDoc to the three.js repository, but adding sample JSDoc for one or two files sooner seems OK to me if that's helpful.

**Comment by Methuselah96** [comment]
Per [this comment](https://github.com/mrdoob/three.js/pull/29357#issuecomment-2345338697), I think we want to steer clear of generating TS definitions or defining types that are specific to TS (e.g., generics/type parameters) in this repo, and focus only on how JSDoc can be used to generate docs. Then in `three-ts-types` I can use the JSDoc as a starting point, and add TS-specific things like generics/type parameters as patches on top of that.

**Comment by DefinitelyMaybe** [comment]
> It would be helpful if someone would be willing to take responsibility for this task.

You called, @WestLangley 😇

#30245 has been labelled as a duplicate of this issue so I shall summarize below.

My contributing idea is simple:

* Introduce a [mono-repo/workspaces](https://docs.npmjs.com/cli/v7/using-npm/workspaces) setup for docs in the repository and
* Adopt [Astro-starlight](https://starlight.astro.build/getting-started/) as three.js's documentation website framework

The migration would not be about changing the look or styling of the docs and `.astro` files are basically a superset of html.

[Astro](https://docs.astro.build/en/getting-started/) itself is successfully running i18n documentation with 14 different languages in-sync with their development. Markdown as a source file format is more approachable for contributors (and maintainers) and starlight has a ton of [tools](https://starlight.astro.build/guides/authoring-content/) for helping content authors.

**Comment by Mugen87** [comment]
At this point, I vote to stop i18n support in the project. `three.js` should focus on providing a auto-generated English documentation via JSDoc. Similar to the maintenance of TS declaration files, the maintenance of translated documentation pages should be managed in a separate repository.

With the migration to JSDoc, we lay the groundwork for new translation workflows. How they will end up looking is something that should be managed elsewhere.

**Comment by DefinitelyMaybe** [comment]
> At this point, I vote to stop i18n support in the project. [...] documentation pages should be managed in a separate repository.

Well if that's what you want, I'm happy to get that started.

**Comment by Mugen87** [comment]
The final decision is up to @mrdoob.

**Comment by DefinitelyMaybe** [comment]
It's actually a good idea btw, Astro contains its docs with a [separate repo](https://github.com/withastro/docs). ya'll could definitely follow suit and offload those concerns/maintenance.

**Comment by Mugen87** [comment]
@mrdoob As a first step towards https://github.com/mrdoob/three.js/issues/24984#issuecomment-2567543726, would you be okay if we remove the languages from the `/docs` directory which are incomplete since more than over a year? That would be `fr`, `jp`, `ko`, `pt-br` and `ru`. All of these are only sparsely maintained.

**Comment by DefinitelyMaybe** [comment]
> Well if that's what you want, I'm happy to get that started.

I created a rough copy of the current docs here:

https://definitelym-threejs-doc-42.deno.dev/

will keep working on it.

**Comment by DefinitelyMaybe** [comment]
Making some good progress on this.

Most of the way through transforming the fundamentals doc content into mdx. There's some content created by scripts and a myriad of iframes (all doing essentially the same thing as `CodeSandbox`, `StackBlitz` or `JSFiddle`) still to go but I'll tackle those later. Is there any plan to unify them or no? I'm assuming no.

**Comment by Mugen87** [comment]
The docs will be replaced with a generated one based on JSDoc. The new docs don't use iFrames.

**Comment by DefinitelyMaybe** [comment]
I've already [generated some docs](https://definitelym-threejs-doc-42.deno.dev/reference/classes/eventdispatcher/) based on the JSDoc's you've worked on.

Planning on generating the rest of those reference docs soon.

> The new docs don't use iFrames.

You mean the generated _reference_ docs, no? afaik, none of the current [reference docs](https://threejs.org/docs/index.html#api/en/animation/AnimationAction) have iframes, I was talking about the [fundamentals doc](https://threejs.org/manual/#en/fundamentals) pages

**Comment by Mugen87** [comment]
> afaik, none of the current [reference docs](https://threejs.org/docs/index.html#api/en/animation/AnimationAction) have iframes,

Yes, they have iFrames. Just check the HTML source.

**Comment by Mugen87** [comment]
We might move such things into the manual instead.

**Comment by DefinitelyMaybe** [comment]
It [already is](https://threejs.org/manual/#en/primitives) but the [materials page](https://threejs.org/manual/#en/materials) isn't.

**Comment by DefinitelyMaybe** [comment]
Ok, I've got as far along with this as reasonable within current params. [Here](https://threejs-docs-six.vercel.app/) is my new deploy.

All current language translations are in. All languages who don't have a translated page fallback to english pages with a page-not-translated warning at the top. The site correctly adjusts for arabic rtl content. The old manual and api reference is there. I also generated a doc using current jsdoc comments. It has search and on-this-page content navigation whose links skip directly to the sub-headings. It does not have examples or properly working inline examples yet. I'd prefer those to become something like [CodeSandbox, StackBlitz or JSFiddle](https://github.com/mrdoob/three.js/issues/24984#issuecomment-2689311142) embeds.

@mrdoob Ready for your input on this one. As far as whats been said, the most popular idea is to move the docs into a separate repo (perhaps [threejs](https://github.com/threejs)?). What it looks like from there may be either a cut and paste of the current docs or potentially an iteration of my demo above.

**Comment by krakowdeveloper** [comment]
I'm trying to make Spanish version without errors, however, I am agree the languages and it's maintenance  could be further block stack, then each language could be labeled as (Beta), leaving the official english language robust.

Some stuff to take in mind: languages are translations from the root docs, but, their external links must keep as original titles in english because some docs (external sources) aren't translated to the relative translation language in mention.

**Comment by Mugen87** [comment]
The old documentation has been eventually replaced with the new JSDoc based, see #30748. Thanks to everyone who worked at the translation over all these years! We have seen remarkable contributions but the way translations were managed/authored in this repository was just not scalable.

**Comment by DefinitelyMaybe** [comment]
Noting comments made [here](https://github.com/mrdoob/three.js/pull/30748#issuecomment-2739770200).

With regards to AI tools, my demo could include an [LLM.txt generator](https://github.com/delucis/starlight-llms-txt) fairly easily.

It sounds like breaking links is an issue maintainers would prefer to avoid. Adopting any new direction, like my demo, would inevitably break that.

I'm might just start a separate project. I'd obviously prefer to contribute towards a [three.js github](https://github.com/threejs) repository. I tried to reach you through discord mrdoob, drop us a line if that's something we could work towards.

**Comment by boytchev** [comment]
Would it be technically feasible for individual docs in a foreign language to have automatically inserted notes/links pointing to the English version if the translation is outdated. In this way users could easily spot if the original English docs are more recent than the translated docs.

**Comment by mrdoob** [comment]
While [trying to improve its performance](https://github.com/jsdoc/jsdoc/pull/2150) I realized that [JSDoc is no longer properly maintained](https://github.com/jsdoc/jsdoc/issues/2129)...

**Comment by Mugen87** [comment]
I would not trust the OP of that issue. Looking at the latest commit history, the project is clearly maintained.

<img width="785" height="748" alt="Image" src="https://github.com/user-attachments/assets/57d1fe69-fdea-48c3-abac-0a3694cd1ac2" />

In general, I would be very reserved with spreading such statements. Given the commit history, that is not fair towards the maintainers of JSDoc, imo. 

If you judge based on open issues, PRs or merged commits per month if projects are maintained or not, a lot of Open-Source projects are in fact "unmaintained".

**Comment by mrdoob** [comment]
Lets see how long it takes for the PR to get reviewed (and hopefully merged) 🤞

**Comment by mrdoob** [comment]
Oh hey! https://github.com/jsdoc/jsdoc/pull/2150#issuecomment-3379369447

**Comment by mrdoob** [comment]
Definitely feeling less anxious about adopting jsdoc now.
